File: app/src/main/java/com/qfq/tainzhi/videoplayer/widget/media/FileMediaDataSource.java
package com.qfq.tainzhi.videoplayer.widget.media;

import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;

import tv.danmaku.ijk.media.player.misc.IMediaDataSource;

/**
 * Created by muqing on 2019/6/1.
 * Email: qfq61@qq.com
 */
public class FileMediaDataSource implements IMediaDataSource {
    private RandomAccessFile mFile;
    private long mFileSize;
    
    public FileMediaDataSource(File file) throws IOException {
        mFile = new RandomAccessFile(file, "r");
        mFileSize = mFile.length();
    }
    
    @Override
    public int readAt(long position, byte[] buffer, int offset, int size) throws IOException {
        if (mFile.getFilePointer() != position)
            mFile.seek(position);
        
        if (size == 0)
            return 0;
        
        return mFile.read(buffer, 0, size);
    }
    
    @Override
    public long getSize() throws IOException {
        return mFileSize;
    }
    
    @Override
    public void close() throws IOException {
        mFileSize = 0;
        mFile.close();
        mFile = null;
    }
}

File: app/src/main/java/com/qfq/tainzhi/videoplayer/media/FileMediaDataSource.kt
package com.qfq.tainzhi.videoplayer.media

import tv.danmaku.ijk.media.player.misc.IMediaDataSource
import java.io.File
import java.io.IOException
import java.io.RandomAccessFile

/**
 * Created by muqing on 2019/6/1.
 * Email: qfq61@qq.com
 */
class FileMediaDataSource constructor(file: File?) : IMediaDataSource {
    private var mFile: RandomAccessFile?
    private var mFileSize: Long
    
    @Throws(IOException::class)
    public override fun readAt(position: Long, buffer: ByteArray?, offset: Int, size: Int): Int {
        if (mFile.getFilePointer() != position) mFile.seek(position)
        if (size == 0) return 0
        return mFile.read(buffer, 0, size)
    }
    
    @Throws(IOException::class)
    public override fun getSize(): Long {
        return mFileSize
    }
    
    @Throws(IOException::class)
    public override fun close() {
        mFileSize = 0
        mFile.close()
        mFile = null
    }
    
    init {
        mFile = RandomAccessFile(file, "r")
        mFileSize = mFile.length()
    }
}